### DISCUSSION TITLE
now() function is considered as untyped by mypy

### DISCUSSION THREAD

**Discussion by zoola969** [sqlalchemy/sqlalchemy#10694] [Usage Questions]
I use mypy 1.6.1 with strict mode, and I get an error when I use `now()` function:
`error: Call to untyped function "now" in typed context  [no-untyped-call]`

### How to reproduce:
`example.py`
```python
from datetime import datetime

from sqlalchemy import DateTime, Integer
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column
from sqlalchemy.sql.functions import now


class Base(DeclarativeBase):
    pass


class Model(Base):
    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=now())

```
`mypy --strict example.py`

How can I avoid marking all places with ` # type: ignore[no-untyped-call]`

**Comment by zoola969** [comment]
same issue with `count` and `sum`

**Reply by zoola969** [reply]
So, I managed to suppress errors with `now()` and `count()`, adding to mypy config:
```toml
[tool.mypy]
untyped_calls_exclude = [
    "sqlalchemy.sql.functions",
]
```
But problem with calls of `sum` and `coalesce` still exists
```python
from sqlalchemy import false
from sqlalchemy.sql.functions import coalesce
from sqlalchemy.sql.functions import count as sa_count

select(
    sa_sum[int](Model.id),
    coalesce[bool](Model.id > 0, false()),
)
```

**Comment by zzzeek** [comment]
hi -

to use SQL functions you use the [func](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.func) namespace as detailed in the [tutorial](https://docs.sqlalchemy.org/en/20/tutorial/data_select.html#tutorial-functions).

```py
from datetime import datetime

from sqlalchemy import DateTime
from sqlalchemy import func
from sqlalchemy import Integer
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column


class Base(DeclarativeBase):
    pass


class Model(Base):
    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
```

**Reply by zoola969** [reply]
What is the purpose of separate `sum`, `coalesce`, etc then?  I thought this is a right way to use typing in sqlachemy.

Please look at 3 statements:
```python
from sqlalchemy import Integer, select
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column
from sqlalchemy.sql.functions import func
from sqlalchemy.sql.functions import sum as sa_sum

class Base(DeclarativeBase):
    pass

class Model(Base):
    id: Mapped[int] = mapped_column(Integer, primary_key=True)

def stmt():
    stmt1 = select(
        func.sum(
            Model.id,
        ).label("sum"),
    ).select_from(
        Model,
    )

    stmt2 = select(
        sa_sum(
            Model.id.label("sum"),
        ).select_from(
            Model,
        ),
    )

    stmt3 = select(
        sa_sum[int](
            Model.id.label("sum"),
        ),
    ).select_from(
        Model,
    )
    reveal_type(stmt1)
    reveal_type(stmt2)
    reveal_type(stmt3)
```

mypy considered types as:
`Revealed type is "sqlalchemy.sql.selectable.Select[Any]"`
`Revealed type is "sqlalchemy.sql.selectable.Select[tuple[Any]]"`
`Revealed type is "sqlalchemy.sql.selectable.Select[tuple[builtins.int]]"`

As you can see, only 3rd statement is correctly typed. And It is important when we want to use mypy with `strict` mode

There is an example:
```python
def typed_select(session: Session) -> int:
    stmt = select(
        sa_sum[int](
            Model.id.label("sum"),
        ),
    ).select_from(
        Model,
    )
    return session.execute(stmt).scalar_one()


def untyped_select(session: Session) -> int:
    stmt = select(
        func.sum(
            Model.id.label("sum"),
        ),
    ).select_from(
        Model,
    )
    return **session.execute(stmt).scalar_one()**
```

In the case of `typed_select` I get `no-untyped-call` error
In the second case I get `no-any-return` error: `Returning Any from function declared to return "int"`

**Reply by zoola969** [reply]
May be we need ability to build such things: `Model.id.sum()`? It will allow to compute type of expression according to type of the model filed.

Or just grant power to `cast` function to override type of expression? Like: `cast(expr, Integer)` -> will be always considered as `int` in the case of type checking

**Reply by zzzeek** [reply]
those classes apply specific behaviors to a selected set of SQL functions that SQLAlchemy chooses to implement.   So they give you desirable runtime behavior, namely the return type at the SQL level being set up correctly which is particularly important for dates and times.   they are invoked automatically when you call them up from func (read the linked tutorial section for background). 

Regarding the typing, the classes themselves are not yet mypy typed and completing this is part of #6810, so when that's completed you probably would not get a mypy error for using the class directly.

**Reply by zzzeek** [reply]
> May be we need ability to build such things: Model.id.sum()? It will allow to compute type of expression according to type of the model filed.

you have this ability, see https://docs.sqlalchemy.org/en/20/core/custom_types.html#redefining-and-creating-new-operators

**Reply by zzzeek** [reply]
> Please look at 3 statements:

This would all work much better when the functions portion of #6810 is done

**Reply by zzzeek** [reply]
we are seeking contributors to continue with #10643 which IMO should be doing the whole module

example (note this was edited, had it wrong on the first run):

```diff
diff --git a/lib/sqlalchemy/sql/functions.py b/lib/sqlalchemy/sql/functions.py
index fc23e9d21..07b6cc3bf 100644
--- a/lib/sqlalchemy/sql/functions.py
+++ b/lib/sqlalchemy/sql/functions.py
@@ -62,12 +62,13 @@ from .. import util
 
 
 if TYPE_CHECKING:
-    from ._typing import _TypeEngineArgument
+    from ._typing import _TypeEngineArgument, _ColumnExpressionArgument
     from ..engine.base import Connection
     from ..engine.cursor import CursorResult
     from ..engine.interfaces import _CoreMultiExecuteParams
     from ..engine.interfaces import CoreExecuteOptionsParameter
 
+
 _T = TypeVar("_T", bound=Any)
 
 _registry: util.defaultdict[
@@ -1048,8 +1049,7 @@ class _FunctionGenerator:
         def session_user(self) -> Type[session_user]:
             ...
 
-        @property
-        def sum(self) -> Type[sum[Any]]:  # noqa: A001
+        def sum(self, *args:_ColumnExpressionArgument[_T], **kw: Any) -> sum[_T]:
             ...
 
         @property
@@ -1387,7 +1387,7 @@ class ReturnTypeFromArgs(GenericFunction[_T]):
 
     inherit_cache = True
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args: _ColumnExpressionArgument[_T], **kwargs: Any):
         fn_args = [
             coercions.expect(
                 roles.ExpressionElementRole,
```

now the above two examples work:

![image](https://github.com/sqlalchemy/sqlalchemy/assets/128223/7dabeb0c-9eee-434f-a5d1-64bcfbedc6c1)

![image](https://github.com/sqlalchemy/sqlalchemy/assets/128223/a0b92450-2b54-4336-bc80-8dbdae7494fb)

**Reply by zzzeek** [reply]
cc @CaselIT

**Reply by CaselIT** [reply]
doing things correctly here does not look easy at all. The correct type here is probably something like `def sum(self, arg: SQLCoreOperations[_T], *args, **kwargs)->sum[_T]` or something along these like (this would be an overload most likely)

**Reply by zzzeek** [reply]
oh right, I missed that.  That's still not hard, ReturnTypeFromArgs stil lhas a fixed signature

here's the correct diff

```diff
diff --git a/lib/sqlalchemy/sql/functions.py b/lib/sqlalchemy/sql/functions.py
index fc23e9d21..07b6cc3bf 100644
--- a/lib/sqlalchemy/sql/functions.py
+++ b/lib/sqlalchemy/sql/functions.py
@@ -62,12 +62,13 @@ from .. import util
 
 
 if TYPE_CHECKING:
-    from ._typing import _TypeEngineArgument
+    from ._typing import _TypeEngineArgument, _ColumnExpressionArgument
     from ..engine.base import Connection
     from ..engine.cursor import CursorResult
     from ..engine.interfaces import _CoreMultiExecuteParams
     from ..engine.interfaces import CoreExecuteOptionsParameter
 
+
 _T = TypeVar("_T", bound=Any)
 
 _registry: util.defaultdict[
@@ -1048,8 +1049,7 @@ class _FunctionGenerator:
         def session_user(self) -> Type[session_user]:
             ...
 
-        @property
-        def sum(self) -> Type[sum[Any]]:  # noqa: A001
+        def sum(self, *args:_ColumnExpressionArgument[_T], **kw: Any) -> sum[_T]:
             ...
 
         @property
@@ -1387,7 +1387,7 @@ class ReturnTypeFromArgs(GenericFunction[_T]):
 
     inherit_cache = True
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args: _ColumnExpressionArgument[_T], **kwargs: Any):
         fn_args = [
             coercions.expect(
                 roles.ExpressionElementRole,

```

output

![Uploading image.png…]()

**Comment by zzzeek** [comment]
I have a merge incoming momentarily
